packages/hash/api/src/model/knowledge/entity.model.ts

```diff
@@ -112,6 +118,145 @@ export default class {
+
+    /**
+     * @todo Once the graph API validates the required links of entities on creation, this may have to be reworked in order
+     *   to create valid entities.
+     *   this code currently creates entities first, then links them together.
+     *   See https://app.asana.com/0/1202805690238892/1203046447168478/f
+     */
+    const entities = await Promise.all(
+      entitiesInTree.map(async (definition) => ({
+        link: definition.meta
+          ? {
+              parentIndex: definition.parentIndex,
+              meta: definition.meta,
+            }
+          : undefined,
+        entity: await EntityModel.getOrCreate(graphApi, {
+          createdById,
+          entityDefinition: definition,
+        }),
+      })),
+    );
+
+    await Promise.all(
+      entities.map(async ({ link, entity }) => {
+        if (link) {
+          const parentEntity = entities[link.parentIndex];
+          if (!parentEntity) {
+            throw new ApolloError("Could not find parent entity");
+          }
+          const linkTypeModel = await LinkTypeModel.get(graphApi, {
+            versionedUri: link.meta.linkTypeVersionedUri,
+          });
+
+          // links are created as an outgoing link from the parent entity to the children.
+          await parentEntity.entity.createOutgoingLink(graphApi, {
+            linkTypeModel,
+            targetEntityModel: entity,
+            index: link.meta.index ?? undefined,
+          });
+        }
+      }),
+    );
+
+    if (entities[0]) {
+      // First element will be the root entity.
+      return entities[0].entity;
+    } else {
+      throw new ApolloError(
+        "Could not create entity tree",
+        "INTERNAL_SERVER_ERROR",
+      );
+    }
+  }
+
+  /**
+   * Get or create an entity given either by new entity properties or a reference
+   * to an existing entity.
+   *
+   * @param params.createdById the account id that is creating the entity
+   * @param params.entityDefinition the definition of how to get or create the entity (excluding any linked entities)
+   */
+  static async getOrCreate(
+    graphApi: GraphApi,
+    params: {
+      createdById: string;
+      entityDefinition: Omit<KnowledgeEntityDefinition, "linkedEntities">;
+    },
+  ): Promise<EntityModel> {
+    const { entityDefinition } = params;
+    const { entityProperties, existingEntity } = entityDefinition;
+
+    let entity;
+
+    if (existingEntity) {
+      entity = await EntityModel.getLatest(graphApi, {
+        entityId: existingEntity.entityId,
+        accountId: existingEntity.ownedById,
+      });
+      if (!entity) {
+        throw new ApolloError(
+          `Entity ${existingEntity.entityId} owned by ${existingEntity.ownedById} not found`,
+          "NOT_FOUND",
+        );
+      }
+    } else if (entityProperties) {
+      const { entityType } = entityDefinition;
+      const { componentId, entityTypeVersionedUri } = entityType ?? {};
```

Comment:
Are we able to have `entityTypeId` here or will that require a bigger change?

Reply:
This is stacked on https://github.com/hashintel/hash/pull/1087, so it should be applied when that is merged and this is rebased

Reply:
Will leave this open then and resolve once we've changed target branch

Reply:
Should be addressed by 5eddccf5d4a25af6b0f4fbc50441003a19569f2e